omegaconf/basecontainer.py

```diff
@@ -301,11 +298,15 @@ def expand(node: Container) -> None:
                 else:
                     node._set_value(type_)
 
-        if dest._is_interpolation() or dest._is_missing():
+        if src._is_missing() and not dest._is_missing():
+            if is_structured_config(get_ref_type(src)):
+                expand(src)
```

Comment:
I think this can have unintended side effects as we may end up instantiating a non-empty object in `src` that would erase some fields.
To see it, run `pytest "tests/structured_conf/test_structured_basic.py::TestStructured::test_merge_structured_onto_dict_nested[tests.structured_conf.data.dataclasses]"` after modifying the class `User` in dataclasses.py with `name: str = "hello"`: notice how the name gets overwritten even though `MissingUserField` is supposed to have a missing `User`.

Reply:
This :
```python
cfg = OmegaConf.merge(
    {"foo": 0, "bar": 1},
    {"foo": "${bar}", "bar": "???"},
)
```

Is as designed because both foo and bar are considered missing.
(By the way, you can merge regular dictionaries directly).

Can you create a minimal example for the second part of your comment?

Reply:
> Is as designed because both foo and bar are considered missing.

Is there anything that relies on `foo` being missing? Under the semantics of "missing == not set", then `foo` wouldn't be missing: it is set to an interpolation pointing to `bar`.
Even if we want to keep considering `foo` as missing, for the specific case of merge, it would make more sense to me to still merge the interpolation. Otherwise it seems confusing that using, e.g. `"foo_${bar}"` would merge the interpolation, while `${bar}` wouldn't. In addition, not being able to merge an interpolation pointing to a missing value seems limiting (for instance, maybe it's because we want to use the value of this variable in the source config).

> (By the way, you can merge regular dictionaries directly).

Oh! :)

> Can you create a minimal example for the second part of your comment?

I added a test on top of master showing exactly what I was alluding to: https://github.com/odelalleau/omegaconf/commit/ca37fd38f4325e88adb016a66867e104eb081917

Reply:
> > Is as designed because both foo and bar are considered missing.
> 
> Is there anything that relies on `foo` being missing? Under the semantics of "missing == not set", then `foo` wouldn't be missing: it is set to an interpolation pointing to `bar`.

Many unit tests at the very least.
The interaction between missing, none, interpolations, structured configs, optional etc is tested well.
If you are asking if we can this is something we can break, I will need a good reason before even considering it.

> Even if we want to keep considering `foo` as missing, for the specific case of merge, it would make more sense to me to still merge the interpolation. Otherwise it seems confusing that using, e.g. `"foo_${bar}"` would merge the interpolation, while `${bar}` wouldn't. In addition, not being able to merge an interpolation pointing to a missing value seems limiting (for instance, maybe it's because we want to use the value of this variable in the source config).

Yes, worth looking if it's easy to not use is_missing here but compare the value directly to MISSING.

> 
> > (By the way, you can merge regular dictionaries directly).
> 
> Oh! :)
> 
> > Can you create a minimal example for the second part of your comment?
> 
> I added a test on top of master showing exactly what I was alluding to: [odelalleau@ca37fd3](https://github.com/odelalleau/omegaconf/commit/ca37fd38f4325e88adb016a66867e104eb081917) (right now, this test fails)

Looking at that test, I am not expecting the name to change after this merge.

Reply:
> If you are asking if we can this is something we can break, I will need a good reason before even considering it.

Followed up on this in your new PR: https://github.com/omry/omegaconf/pull/464#issuecomment-748111885

> > I added a test on top of master showing exactly what I was alluding to: [odelalleau@ca37fd3](https://github.com/odelalleau/omegaconf/commit/ca37fd38f4325e88adb016a66867e104eb081917) (right now, this test fails)
> 
> Looking at that test, I am not expecting the name to change after this merge.

Currently it does, and my guess is it's because of the line I pointed to. A potential fix (not sure if there's a better way) would be to set all fields of `src` to `MISSING` after the expand.

Reply:
Ouch.
Need to take a closer look at it. switching. can you file a standalone bug to keep track of it?

Reply:
> Need to take a closer look at it. Can you file a standalone bug to keep track of it?

I created a PR with a fix instead to (hopefully) save time: #465